pkg/gardenlet/controller/bastion/bastion.go

```diff
@@ -0,0 +1,164 @@
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+	"github.com/gardener/gardener/pkg/logger"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	runtimecache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Controller controls BackupEntries.
+type Controller struct {
+	gardenClient client.Client
+	config       *config.GardenletConfiguration
+	reconciler   reconcile.Reconciler
+	recorder     record.EventRecorder
+
+	bastionInformer runtimecache.Informer
+	bastionQueue    workqueue.RateLimitingInterface
+
+	workerCh               chan int
+	numberOfRunningWorkers int
+}
+
+// NewBastionController takes a Kubernetes client for the Garden clusters <k8sGardenClient>, a struct
+// holding information about the acting Gardener, a <bastionInformer>, and a <recorder> for
+// event recording. It creates a new Gardener controller.
+func NewBastionController(ctx context.Context, clientMap clientmap.ClientMap, config *config.GardenletConfiguration, recorder record.EventRecorder) (*Controller, error) {
+	gardenClient, err := clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get garden client: %w", err)
+	}
+
+	bastionInformer, err := gardenClient.Cache().GetInformer(ctx, &gardencorev1alpha1.Bastion{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Bastion Informer: %w", err)
+	}
+
+	return &Controller{
+		gardenClient:    gardenClient.Client(),
+		config:          config,
+		reconciler:      newReconciler(clientMap, recorder, config),
+		recorder:        recorder,
+		bastionInformer: bastionInformer,
+		bastionQueue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Bastion"),
+		workerCh:        make(chan int),
+	}, nil
+}
+
+// Run runs the Controller until the given stop channel can be read from.
+func (c *Controller) Run(ctx context.Context, workers int) {
+	var waitGroup sync.WaitGroup
+
+	c.bastionInformer.AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controllerutils.BastionFilterFunc(ctx, c.gardenClient, confighelper.SeedNameFromSeedConfig(c.config.SeedConfig), c.config.SeedSelector),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.bastionAdd,
+			UpdateFunc: c.bastionUpdate,
+			DeleteFunc: c.bastionDelete,
+		},
+	})
```

Comment:
Can you move this up into the `NewBastionController` function? I see that you were inspired from the `BackupEntry` controllers and I have no explanation why it was done like this, but I guess most controllers register the event handlers already in the `New()` functions. Maybe you can also move it up in both the `BackupBucket`/`BackupEntry` controllers